File: auth/internal/domain/service/seeding_service.go
package service

import (
	"context"

	"tafu-auth/internal/domain/entity"
	"tafu-auth/internal/domain/repository"
	"tafu-auth/pkg/logger"
)

type SeedingService struct {
	roleRepo           repository.RoleRepository
	permissionRepo     repository.PermissionRepository
	rolePermissionRepo repository.RolePermissionRepository
}

func NewSeedingService(
	roleRepo repository.RoleRepository,
	permissionRepo repository.PermissionRepository,
	rolePermissionRepo repository.RolePermissionRepository,
) *SeedingService {
	return &SeedingService{
		roleRepo:           roleRepo,
		permissionRepo:     permissionRepo,
		rolePermissionRepo: rolePermissionRepo,
	}
}

func (s *SeedingService) SeedAll(ctx context.Context) error {
	logger.Info().Msg("Starting database seeding...")

	if err := s.SeedRoles(ctx); err != nil {
		return err
	}

	if err := s.SeedPermissions(ctx); err != nil {
		return err
	}

	if err := s.SeedRolePermissions(ctx); err != nil {
		return err
	}

	logger.Info().Msg("Database seeding completed")
	return nil
}

func (s *SeedingService) SeedRoles(ctx context.Context) error {
	logger.Info().Msg("Seeding roles...")

	for _, role := range entity.DefaultRoles {
		if err := s.roleRepo.Upsert(ctx, &role); err != nil {
			logger.Error().Err(err).Str("role", role.Name).Msg("Failed to seed role")
			return err
		}
	}

	logger.Info().Int("count", len(entity.DefaultRoles)).Msg("Roles seeded successfully")
	return nil
}

func (s *SeedingService) SeedPermissions(ctx context.Context) error {
	logger.Info().Msg("Seeding permissions...")

	for _, permission := range entity.DefaultPermissions {
		if err := s.permissionRepo.Upsert(ctx, &permission); err != nil {
			logger.Error().Err(err).Str("permission", permission.Slug).Msg("Failed to seed permission")
			return err
		}
	}

	logger.Info().Int("count", len(entity.DefaultPermissions)).Msg("Permissions seeded successfully")
	return nil
}

func (s *SeedingService) SeedRolePermissions(ctx context.Context) error {
	logger.Info().Msg("Seeding role-permission mappings...")

	for roleName, permissionSlugs := range entity.RolePermissionMapping {
		// Skip SUPER_ADMIN as it has wildcard permissions
		if roleName == entity.RoleSuperAdmin {
			continue
		}

		role, err := s.roleRepo.FindByName(ctx, roleName)
		if err != nil {
			logger.Warn().Err(err).Str("role", roleName).Msg("Role not found, skipping")
			continue
		}

		for _, slug := range permissionSlugs {
			permission, err := s.permissionRepo.FindBySlug(ctx, slug)
			if err != nil {
				logger.Warn().Err(err).Str("permission", slug).Msg("Permission not found, skipping")
				continue
			}

			// Check if mapping exists
			exists, err := s.rolePermissionRepo.Exists(ctx, role.ID, permission.ID)
			if err != nil {
				logger.Warn().Err(err).Msg("Failed to check role-permission existence")
				continue
			}

			if !exists {
				rp := &entity.RolePermission{
					RoleID:       role.ID,
					PermissionID: permission.ID,
				}
				if err := s.rolePermissionRepo.Create(ctx, rp); err != nil {
					logger.Warn().Err(err).
						Str("role", roleName).
						Str("permission", slug).
						Msg("Failed to create role-permission mapping")
				}
			}
		}
	}

	logger.Info().Msg("Role-permission mappings seeded successfully")
	return nil
}

func (s *SeedingService) ResyncRolePermissions(ctx context.Context, roleName string) error {
	role, err := s.roleRepo.FindByName(ctx, roleName)
	if err != nil {
		return err
	}

	// Delete existing permissions
	if err := s.rolePermissionRepo.DeleteByRoleID(ctx, role.ID); err != nil {
		return err
	}

	// Re-add permissions
	permissionSlugs, ok := entity.RolePermissionMapping[roleName]
	if !ok {
		return nil
	}

	for _, slug := range permissionSlugs {
		permission, err := s.permissionRepo.FindBySlug(ctx, slug)
		if err != nil {
			continue
		}

		rp := &entity.RolePermission{
			RoleID:       role.ID,
			PermissionID: permission.ID,
		}
		if err := s.rolePermissionRepo.Create(ctx, rp); err != nil {
			logger.Warn().Err(err).
				Str("role", roleName).
				Str("permission", slug).
				Msg("Failed to create role-permission mapping")
		}
	}

	return nil
}

func (s *SeedingService) ResetAllSeededData(ctx context.Context) error {
	logger.Warn().Msg("Resetting all seeded data...")

	// Delete role permissions first (foreign key constraint)
	for _, role := range entity.DefaultRoles {
		r, err := s.roleRepo.FindByName(ctx, role.Name)
		if err != nil {
			continue
		}
		if err := s.rolePermissionRepo.DeleteByRoleID(ctx, r.ID); err != nil {
			logger.Warn().Err(err).Str("role", role.Name).Msg("Failed to delete role permissions")
		}
	}

	// Delete permissions
	for _, permission := range entity.DefaultPermissions {
		p, err := s.permissionRepo.FindBySlug(ctx, permission.Slug)
		if err != nil {
			continue
		}
		if err := s.permissionRepo.Delete(ctx, p.ID); err != nil {
			logger.Warn().Err(err).Str("permission", permission.Slug).Msg("Failed to delete permission")
		}
	}

	// Delete roles
	for _, role := range entity.DefaultRoles {
		r, err := s.roleRepo.FindByName(ctx, role.Name)
		if err != nil {
			continue
		}
		if err := s.roleRepo.Delete(ctx, r.ID); err != nil {
			logger.Warn().Err(err).Str("role", role.Name).Msg("Failed to delete role")
		}
	}

	logger.Info().Msg("Seeded data reset completed")
	return nil
}

File: auth-profile-nestjs/auth/src/modules/identity/services/seeding.service.ts
import { Injectable, Logger, OnModuleInit } from '@nestjs/common';
import { InjectRepository } from '@nestjs/typeorm';
import { Repository } from 'typeorm';

import { Role, Permission, RolePermission } from '../entities';
import {
  ROLE_DEFINITIONS,
  PERMISSION_DEFINITIONS,
  ROLE_PERMISSION_MAPPING,
  DEFAULT_ROLES,
} from '../constants';

@Injectable()
export class SeedingService implements OnModuleInit {
  private readonly logger = new Logger(SeedingService.name);

  constructor(
    @InjectRepository(Role)
    private readonly roleRepository: Repository<Role>,
    @InjectRepository(Permission)
    private readonly permissionRepository: Repository<Permission>,
    @InjectRepository(RolePermission)
    private readonly rolePermissionRepository: Repository<RolePermission>,
  ) {}

  async onModuleInit(): Promise<void> {
    await this.seed();
  }

  async seed(): Promise<void> {
    this.logger.log('Starting database seeding...');

    try {
      await this.seedPermissions();
      await this.seedRoles();
      await this.seedRolePermissions();

      this.logger.log('Database seeding completed successfully');
    } catch (error) {
      this.logger.error(
        'Database seeding failed',
        error instanceof Error ? error.stack : String(error),
      );
    }
  }

  private async seedPermissions(): Promise<void> {
    let created = 0;
    let skipped = 0;

    for (const perm of PERMISSION_DEFINITIONS) {
      const slug = `${perm.resource}:${perm.action}`;
      const exists = await this.permissionRepository.findOne({
        where: { slug },
      });

      if (!exists) {
        const permission = this.permissionRepository.create({
          resource: perm.resource,
          action: perm.action,
          slug,
          displayName: perm.displayName,
          category: perm.category,
          isDangerous: 'isDangerous' in perm ? perm.isDangerous : false,
          description: `Permission to ${perm.action} ${perm.resource}`,
        });
        await this.permissionRepository.save(permission);
        created++;
      } else {
        skipped++;
      }
    }

    this.logger.log(`Permissions: ${created} created, ${skipped} already exist`);
  }

  private async seedRoles(): Promise<void> {
    let created = 0;
    let skipped = 0;

    for (const roleData of ROLE_DEFINITIONS) {
      const exists = await this.roleRepository.findOne({
        where: { name: roleData.name },
      });

      if (!exists) {
        const role = this.roleRepository.create({
          name: roleData.name,
          displayName: roleData.displayName,
          description: `${roleData.displayName} role`,
          isSystem: roleData.isSystem,
          isDefault: 'isDefault' in roleData ? roleData.isDefault : false,
          priority: roleData.priority,
        });
        await this.roleRepository.save(role);
        created++;
      } else {
        skipped++;
      }
    }

    this.logger.log(`Roles: ${created} created, ${skipped} already exist`);
  }

  private async seedRolePermissions(): Promise<void> {
    let created = 0;
    let skipped = 0;

    for (const [roleName, permissionSlugs] of Object.entries(ROLE_PERMISSION_MAPPING)) {
      const role = await this.roleRepository.findOne({
        where: { name: roleName },
      });

      if (!role) {
        this.logger.warn(`Role not found: ${roleName}`);
        continue;
      }

      // Handle super admin with all permissions
      if (roleName === DEFAULT_ROLES.SUPER_ADMIN && permissionSlugs.includes('*')) {
        const allPermissions = await this.permissionRepository.find();
        for (const permission of allPermissions) {
          const exists = await this.rolePermissionRepository.findOne({
            where: { roleId: role.id, permissionId: permission.id },
          });

          if (!exists) {
            const rolePermission = this.rolePermissionRepository.create({
              roleId: role.id,
              permissionId: permission.id,
            });
            await this.rolePermissionRepository.save(rolePermission);
            created++;
          } else {
            skipped++;
          }
        }
        continue;
      }

      for (const slug of permissionSlugs) {
        if (slug === '*') continue;

        const permission = await this.permissionRepository.findOne({
          where: { slug },
        });

        if (!permission) {
          this.logger.warn(`Permission not found: ${slug}`);
          continue;
        }

        const exists = await this.rolePermissionRepository.findOne({
          where: { roleId: role.id, permissionId: permission.id },
        });

        if (!exists) {
          const rolePermission = this.rolePermissionRepository.create({
            roleId: role.id,
            permissionId: permission.id,
          });
          await this.rolePermissionRepository.save(rolePermission);
          created++;
        } else {
          skipped++;
        }
      }
    }

    this.logger.log(`Role-Permissions: ${created} created, ${skipped} already exist`);
  }

  /**
   * Reset all seeded data (for testing)
   */
  async reset(): Promise<void> {
    this.logger.warn('Resetting seeded data...');

    await this.rolePermissionRepository.delete({});
    await this.permissionRepository.delete({});
    await this.roleRepository.delete({ isSystem: true });

    this.logger.log('Seeded data reset complete');
  }

  /**
   * Resync permissions for a specific role
   */
  async resyncRolePermissions(roleName: string): Promise<void> {
    const role = await this.roleRepository.findOne({
      where: { name: roleName },
    });

    if (!role) {
      throw new Error(`Role not found: ${roleName}`);
    }

    const permissionSlugs = ROLE_PERMISSION_MAPPING[roleName];
    if (!permissionSlugs) {
      throw new Error(`No permission mapping found for role: ${roleName}`);
    }

    // Remove existing permissions
    await this.rolePermissionRepository.delete({ roleId: role.id });

    // Re-add permissions
    for (const slug of permissionSlugs) {
      if (slug === '*') {
        const allPermissions = await this.permissionRepository.find();
        for (const permission of allPermissions) {
          const rolePermission = this.rolePermissionRepository.create({
            roleId: role.id,
            permissionId: permission.id,
          });
          await this.rolePermissionRepository.save(rolePermission);
        }
      } else {
        const permission = await this.permissionRepository.findOne({
          where: { slug },
        });

        if (permission) {
          const rolePermission = this.rolePermissionRepository.create({
            roleId: role.id,
            permissionId: permission.id,
          });
          await this.rolePermissionRepository.save(rolePermission);
        }
      }
    }

    this.logger.log(`Resynced permissions for role: ${roleName}`);
  }
}
